server/athenian/api/controllers/search_controller/search_prs.py

```diff
@@ -220,3 +236,97 @@ async def _search_pr_digests(
             ",".join(map(str, unknown_prs)),
         )
     return pr_digests
+
+
+@sentry_span
+def _apply_order_by(
+    pr_facts: pd.DataFrame,
+    search_filter: _SearchPRsFilter,
+    order_by: Sequence[SearchPullRequestsOrderByExpression],
+) -> pd.DataFrame:
+    assert order_by
+    if not len(pr_facts):
+        return pr_facts
+
+    keep_mask = np.full((len(pr_facts),), True, bool)
+    ordered_indexes = np.arange(len(pr_facts))
+    order_by_metrics: Optional[_OrderByMetrics] = None
+
+    for expr in reversed(order_by):
+        if expr.field in _OrderByMetrics.FIELDS:
+            if order_by_metrics is None:
+                order_by_metrics = _OrderByMetrics.build(pr_facts, search_filter, order_by)
+            ordered_indexes, discard = order_by_metrics.parse_expression(expr, ordered_indexes)
+            keep_mask[discard] = False
+
+    kept_positions = ordered_indexes[np.nonzero(keep_mask[ordered_indexes])]
+    return pr_facts.iloc[kept_positions]
+
+
+class _OrderByMetrics:
+    """Handles order by pull request metric values.
+
+    Expressions about a metric field can be fed into this object with `parse_expression`.
+    """
+
+    FIELDS = pr_metric_calculators
+
+    def __init__(self, calc_ensemble: PullRequestMetricCalculatorEnsemble):
+        self._calc_ensemble = calc_ensemble
+
+    @classmethod
+    def build(
+        cls,
+        pr_facts: pd.DataFrame,
+        search_filter: _SearchPRsFilter,
+        order_by: Sequence[SearchPullRequestsOrderByExpression],
+    ) -> _OrderByMetrics:
+        metrics = [expr.field for expr in order_by if expr.field in cls.FIELDS]
+        assert metrics
+        min_times, max_times = (
+            np.array([t.replace(tzinfo=None)], dtype="datetime64[ns]")
+            for t in (search_filter.time_from, search_filter.time_to)
+        )
+        calc_ensemble = PullRequestMetricCalculatorEnsemble(
+            *metrics, quantiles=(0, 1), quantile_stride=DEFAULT_QUANTILE_STRIDE,
+        )
+        groups = np.full((1, len(pr_facts)), True, bool)
+        calc_ensemble(pr_facts, min_times, max_times, groups)
+        return cls(calc_ensemble)
+
+    def parse_expression(
```

Comment:
I'd rather rename this to `parse_and_apply_expression` since we are doing quite some extra work besides parsing.